### PR TITLE
Packages page

### PR DESCRIPTION
**Pull request by daveverwer**
As discussed, here’s a first draft of a pull request to showcase some of the Swift package ecosystem being tracked by the Swift Package Index on Swift.org.

This is not ready to merge, but @finestructure and I want to open it early as it’s the start of a significant addition to the site.

### Demo

[View a demo/staging version of this pull request](https://spi-swift-org.netlify.app/packages/)

### Package Data

If merged, this PR has an initial data set in `_data/packages/packages.yml` that will need regular updating. We have a tool in development that will automate the updates on a schedule to be agreed upon (we’d suggest every two weeks is appropriate). This tool will submit all data set updates via pull request to this repository.

**Note:** This draft PR does not include up-to-date data. The data in this initial draft is real but was hand-written and outdated. Please treat it as representative rather than final. We will update this PR with fresh data as soon as the tool can generate it.

#### Community Showcase

This list of packages should be the most dynamic and will change most often. Initially, this list will consist of packages discussed on [Swift Package Indexing](https://swiftpackageindexing.transistor.fm/), but with community outreach will expand to more sources.

#### Keyword Data

The current keywords are `networking` and `testing`. We may want to add more or change these. The packages picked for each keyword are ranked by the [SPI internal search ranking score](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/blob/42cac0e3c00dd68a3b5bd587f8bc445026f43614/Sources/App/Core/Score.swift). The packages and order can change over time, but we’d expect changes to be gradual.

### To Do

- [x] Decide on an initial set of keywords and a process to decide if/when those keywords should change or be reviewed.
- [x] Replace `icon-external-color.svg`. It is currently a modified version of an icon from the Noun Project. This needs re-drawing or licensing.
- [x] ~~We may also need to get approval for the OSI logo in `assets/images/icon-osi.svg`.~~ Removed this icon for now, it's too small in this position to comply with their [logo guidelines](https://opensource.org/logo-usage-guidelines/).
- [x] Refine package list design on mobile devices.
- [x] Reach out to some podcasts, sites, and newsletters in the community to gather sources for community showcase packages.
- [x] Update the Community Showcase preamble with the link to the forum post.
- [x] Final update of the packages from latest SPI data

### CODE REVIEW

File: .vscode/tasks.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+      "presentation": {
+        "reveal": "always"
+      },
+      "group": {
+        "kind": "build",
+        "isDefault": true
+      }
+    }
+  ]
+}
```
Comment (tomerd) [inline .vscode/tasks.json:19]: redundant?

File: packages/index.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+---
+layout: page
+title: Packages
+---
+
+Swift has a rich ecosystem of open-source packages that you can easily import into your project using the [Swift Package Manager]({% link package-manager/index.md %}) (SPM). With thousands of open-source packages, you’re sure to find what you’re looking for, whether it’s a package to help you build great apps for Apple platforms or if you’re looking to develop server backends with Swift. Or, get started today by [creating a package yourself](https://developer.apple.com/documentation/xcode/creating-a-standalone-swift-package-with-xcode).
```
Comment (tomerd) [inline packages/index.md:6]: we normally use `SwiftPM` over `SPM`

**Comment by daveverwer** [comment]
> @daveverwer could you add some screenshots or (even better) a link to some live version of this? it will make reviewing the changes and providing UI/UX feedback easier

There is a demo link in the opening description. I've moved it to the top and it's now the first heading.

File: packages/index.md (hunk 2 above)
Comment (daveverwer) [reply packages/index.md:6]: Fixed. Thanks!

File: .vscode/tasks.json (hunk 1 above)
Comment (daveverwer) [reply .vscode/tasks.json:19]: Apologies, this was an accidental commit of a file that made working with the project easier. Removed.

**Comment by daveverwer** [comment]
Quick update.

This PR and [hosted prototype](https://spi-swift-org.netlify.app/packages/) now includes real data produced by [our tool](https://github.com/SwiftPackageIndex/PackageListTool) and an additional section, based on an idea we had to highlight new language features. It highlights the top 6 packages on the SPI that include macro products.

**Review by 0xTim — APPROVED** [review]
I think this is a great addition to the site! We'll need to be conscious of ensuring that the PRs with updated data don't get stale (is there an SPI API we could pull it from?) and it would be good to have a guide on creating packages on Swift.org itself but overall this looks great!

**Comment by daveverwer** [comment]
> We'll need to be conscious of ensuring that the PRs with updated data don't get stale (is there an SPI API we could pull it from?)

My hope is that these PRs can be fast tracked since they will be in the same format every time. Just data updates. We should discuss this at the next SWWG meeting.

It would be possible to pull from the SPI API for some categories during site build, meaning the data in some categories would always be up to date. That would not be ideal for packages in the Community Showcase, though, as that needs manual curation.

My view is that it's best to do all updates in the same way, but we can also discuss that. Either here or at the WG meeting.

**Comment by 0xTim** [comment]
> My view is that it's best to do all updates in the same way, but we can also discuss that. Either here or at the WG meeting.

Was this discussed as I missed the meeting?

Either way this is a massive improvement and further improves swift.org! In the interests of automating things, if you're building a tool to create the PRs could the tool not expose those packages via an API that we then load here via JS? There's probably a can of worms there but worth raising at least as something to consider long term

**Comment by daveverwer** [comment]
> Was this discussed as I missed the meeting?
> 
> Either way this is a massive improvement and further improves swift.org! In the interests of automating things, if you're building a tool to create the PRs could the tool not expose those packages via an API that we then load here via JS? There's probably a can of worms there but worth raising at least as something to consider long term

Unfortunately I was also unable to attend the meeting and so I'm not sure if it was discussed. I don't think it was.

The tool currently is a command line tool that we will run manually at first, and potentially in an automated way in the future rather than being a hosted service. [It's open-source if you'd like to see it](https://github.com/SwiftPackageIndex/PackageListTool), but just a note that it's not yet complete!

**Review by alexandersandberg — COMMENTED** [review]
I think it would be nice to make this "Explore more..." link more prominent somehow. Currently, it's easy to miss it when quickly scrolling through and scanning the list, making it almost appear that these are _all_ the packages that exist, not just a small selection.

CleanShot 2023-08-07 at 11 14 44

> Unfortunately I was also unable to attend the meeting and so I'm not sure if it was discussed. I don't think it was.

This was not discussed in the last meeting.

**Comment by daveverwer** [comment]
> I think it would be nice to make this "Explore more..." link more prominent somehow. Currently, it's easy to miss it when quickly scrolling through and scanning the list, making it almost appear that these are _all_ the packages that exist, not just a small selection.

This is a good idea. Maybe one of these "Learn More" buttons with "More {{something}} Packages"?

Screenshot 2023-08-07 at 10 37 42@2x

I think that's about the closest we have to a larger action button design. I could either re-create those styles in this PR and remove them once we get `content-improvements` merged, or we could change this CTA as part of the `content-improvements` branch.

**Comment by alexandersandberg** [comment]
> This is a good idea. Maybe one of these "Learn More" buttons with "More {{something}} Packages"?
> 
> Screenshot 2023-08-07 at 10 37 42@2x
> 
> I think that's about the closest we have to a larger action button design. I could either re-create those styles in this PR and remove them once we get `content-improvements` merged, or we could change this CTA as part of the `content-improvements` branch.

I think just copy-pasting the new styles into this branch is fine, and then maybe adding a TODO item for the final content improvements PR to remove those.

**Comment by daveverwer** [comment]
> I think just copy-pasting the new styles into this branch is fine, and then maybe adding a TODO item for the final content improvements PR to remove those.

I didn't see this until just now but I ended up going the other way. I shortened the CTAs and used the `›` character that we're using elsewhere in the `content-improvements` branch:

Screenshot 2023-08-07 at 11 32 18@2x

**Comment by alexandersandberg** [comment]
> I didn't see this until just now but I ended up going the other way. I shortened the CTAs and used the `›` character that we're using elsewhere in the `content-improvements` branch:
> 
> Screenshot 2023-08-07 at 11 32 18@2x

That's an improvement! I still feel like it could be more prominent though. Maybe adjusting the spacing between the sections could help with that?

File: _data/packages/packages.yml [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+          for working with Swift macros. It simplifies tasks such as parsing floating
+          point literals, type destructuring, type normalization, and getting the value
+          of string literals.
+        swift_compatibility: 5.9+
+        platform_compatibility:
+          - Apple
+        platform_compatibility_tooltip: Apple (macOS)
+        license: Apache 2.0
+        url: https://swiftpackageindex.com/stackotter/swift-macro-toolkit
+  - name: Swift on the Server
```
Comment (cthielen) [inline _data/packages/packages.yml:160]: @tomerd Do we have a convention for referring to server use cases? I feel like we've been using "Swift on server" (with that casing) lately?

File: _data/packages/packages.yml [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+    anchor: server
+    description: What packages can help you build a top-class back-end server for your
+      app or a website built with HTML? Swift has several mature back-end frameworks,
+      ORM and database connection packages, and more.
```
Comment (cthielen) [inline _data/packages/packages.yml:164]: Recommend "ORM, database connection packages, and more"

File: _data/packages/packages.yml [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+    more:
+      title: "More Swift on the Server packages \u203A"
```
Comment (cthielen) [inline _data/packages/packages.yml:166]: If we change the heading, e.g. to "Swift on server", let's update it here as well

File: packages/index.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
+---
+layout: page
+title: Packages
+---
+
+Swift has a rich ecosystem of open-source packages that you can easily import into your project using the [Swift Package Manager]({% link package-manager/index.md %}) (SwiftPM). With thousands of open-source packages, you’re sure to find what you’re looking for, whether it’s a package to help you build great apps for Apple platforms or if you’re looking to develop server backends with Swift. Or, get started today by [creating a package yourself](https://developer.apple.com/documentation/xcode/creating-a-standalone-swift-package-with-xcode).
```
Comment (cthielen) [inline packages/index.md:6]: I believe we try to use "open source" instead of "open-source".

Comment (cthielen) [reply packages/index.md:6]: The second sentence feels a little over-the-top. How about:

"
Swift has a rich ecosystem of open source packages that you can easily import into your project using the [Swift Package Manager]({% link package-manager/index.md %}) (SwiftPM).

There are thousands of packages to help with networking, testing, animation, (or whichever categories make the most sense).

You can also [create your own package](https://developer.apple.com/documentation/xcode/creating-a-standalone-swift-package-with-xcode) for personal use or to share with others.
"

File: _data/packages/packages.yml (hunk 3 above)
Comment (daveverwer) [reply _data/packages/packages.yml:160]: The capitalisation throughout Swift.org for this term varies, but "Swift on server" and "Swift on Server" is about twice as common than "Swift on the Server" and "Server-Side Swift".

I've changed this to "Swift on server" for now.

File: packages/index.md (hunk 6 above)
Comment (daveverwer) [reply packages/index.md:6]: Fixed 👍

File: _data/packages/packages.yml (hunk 4 above)
Comment (daveverwer) [reply _data/packages/packages.yml]: Fixed

File: _data/packages/packages.yml (hunk 3 above)
Comment (tomerd) [reply _data/packages/packages.yml]: _Swift on server_ is my personal preference

File: index.md [hunk 7]
```diff
@@ -41,6 +41,13 @@ atom: true
         Read the docs
       </a>
     </li>
+
+    <li>
+      <a href="/packages">
+        <img src="/assets/images/landing-page/box.svg" />
+        Explore Packages
```
Comment (alexandersandberg) [inline index.md:48]: Maybe lowercase "packages" to keep it consistent with the other boxes?

```suggestion
        Explore packages
```

Currently:
CleanShot 2023-09-27 at 07 55 30

Comment (daveverwer) [reply index.md:48]: This came up during last night's meeting. Already fixed locally, I just didn't push yet.

**Comment by daveverwer** [comment]
Hi All. I just pushed a version that implements all the feedback from Tuesday’s meeting.

I haven’t cleaned up the underlying code yet as I didn’t want to do that until I have broad agreement that this structure works for everyone.

Please leave all comments that cover layout, structure, or content. I’m happy with how this looks and behaves, please leave any feedback below.

The preview here: https://spi-swift-org.netlify.app/packages/ is updated with all these changes.

**Comment by 0xTim** [comment]
I agree with you that it's a shame not to have Community in the new and notable header. My only comment is we could probably link to the creating a library tutorial we have on the site rather than one on the Apple site. However none of these are blockers and I vote merge and fix forward!

**Comment by daveverwer** [comment]
> I agree with you that it's a shame not to have Community in the new and notable header.

I still like Community Showcase of all the options put forward.

> My only comment is we could probably link to the creating a library tutorial we have on the site rather than one on the Apple site.

Fixed in 952e97cd30fab642bb30b149a57664c33fbff2d5

**Comment by 0xTim** [comment]
> I still like Community Showcase of all the options put forward.

+1 from me on that

**Comment by daveverwer** [comment]
I've now cleaned up the CSS from the previous experiments, and so this should be good for another full review.

**Comment by kaishin** [comment]
How about "Community Spotlight" as a title for the first category? I personally prefer that to both "Community Showcase" and the current "New and Notable".

In a later iteration we can even give it a different visual treatment to make it stand out more.

File: packages/index.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+## Package Ecosystem
+
+Browse a small selection of interesting packages in popular categories from around the community, as well as a hand-picked selection of new and notable packages.
+
+{% include_relative _package-lists.html %}
+
+## Create Your Own
+
+Creating a Swift package is a great way to modularise your code either for personal use, for use privately inside your company, or if you want to release an open source package for the rest of the Swift community to use. Start by reading [creating a Swift package](https://www.swift.org/getting-started/library-swiftpm/), or by watching [Creating Swift Packages](https://developer.apple.com/videos/play/wwdc2019/410/) from WWDC 2019, or [Meet Swift Package plugins](https://developer.apple.com/videos/play/wwdc2022/110359/) from WWDC 2022.
```
Comment (shahmishal) [inline packages/index.md:16]:
```suggestion
Creating a Swift package is a great way to modularise your code either for personal use, for use privately inside your company, or if you want to release an open source package for the rest of the Swift community to use. Start by reading [creating a Swift package](/getting-started/library-swiftpm/), or by watching [Creating Swift Packages](https://developer.apple.com/videos/play/wwdc2019/410/) from WWDC 2019, or [Meet Swift Package plugins](https://developer.apple.com/videos/play/wwdc2022/110359/) from WWDC 2022.
```

Comment (daveverwer) [reply packages/index.md]: Thanks again!

File: _data/packages/packages.yml [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+        owner: Jon Reid
+        swift_compatibility: 5.6+
+        platform_compatibility:
+          - Apple
+        platform_compatibility_tooltip: Apple (iOS, macOS, visionOS, watchOS, tvOS)
+        license: MIT
+        url: https://swiftpackageindex.com/jonreid/OCMockito
+  - name: Debug logging
+    slug: logging
+    brief: Are you looking for something more than Swift's built-in logging? Start with
```
Comment (dempseyatgithub) [inline _data/packages/packages.yml:436]: smart quote:

Should be “Swift’s”

File: _data/packages/packages.yml [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+        platform_compatibility:
+          - Apple
+          - Linux
+        platform_compatibility_tooltip: Apple (iOS, macOS, visionOS, watchOS, tvOS) and
+          Linux
+        license: Apache 2.0
+        url: https://swiftpackageindex.com/Kitura/BlueSocket
+  - name: Testing
+    slug: testing
+    brief: If you want to level up your project's tests, take a look at a selection
```
Comment (dempseyatgithub) [inline _data/packages/packages.yml:353]: smart quote:

Should be “project’s”

File: _data/packages/packages.yml [hunk 11]
```diff
@@ -0,0 +1,508 @@
+categories:
+  - name: Community Showcase
+    slug: showcase
+    brief: A hand-curated selection of new and interesting packages being discussed
+      by the Swift community. Browse the list to find out what's new and how to nominate
```
Comment (dempseyatgithub) [inline _data/packages/packages.yml:5]: smart quote:

Should be “what’s”

(Jekyll apparently doesn’t always give front matter smart quotes?)

File: packages/index.md [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+## Package Ecosystem
+
+Browse a small selection of interesting packages in popular categories from around the community, as well as a hand-picked selection of new and notable packages.
+
+{% include_relative _package-lists.html %}
+
+## Create Your Own
+
+Creating a Swift package is a great way to modularise your code either for personal use, for use privately inside your company, or if you want to release an open source package for the rest of the Swift community to use. Start by reading [creating a Swift package](/getting-started/library-swiftpm/), or by watching [Creating Swift Packages](https://developer.apple.com/videos/play/wwdc2019/410/) from WWDC 2019, or [Meet Swift Package plugins](https://developer.apple.com/videos/play/wwdc2022/110359/) from WWDC 2022.
```
Comment (dempseyatgithub) [inline packages/index.md:16]: The three options in this paragraph each use a different structure. Maybe something like:

“either for personal use, for private use inside your company, or to release an open source package...”

Also, possibly make ‘Start by reading ...” a new paragraph. Doing so gives additional weight to the call to action by not burying the first recommended step in the middle of a paragraph.

File: packages/index.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+Then, if you'd like to let other members of the community use your open source package, [add it to the Swift Package Index](https://swiftpackageindex.com/add-a-package).
+
+## Next Steps
+
+The package lists above show a selection of the broader Swift package ecosystem. The [Swift Package Index](https://swiftpackageindex.com/) is a community-run search engine for Swift packages with powerful filters to help you find the packages that work for your project. For example, you could search for packages that:
+
+* [Match a search term](https://swiftpackageindex.com/search?query=swiftui)
+* [Have compatibility with both iOS and macOS platforms](https://swiftpackageindex.com/search?query=swiftui+platform:ios,macos)
+* [Include an executable helper or tool](https://swiftpackageindex.com/search?query=swiftui+product:executable)
+* [Have been created by a specific author](https://swiftpackageindex.com/search?query=author:apple)
```
Comment (dempseyatgithub) [inline packages/index.md:27]: It feels odd to end the page on this list. Possibly add a final sentence that sums everything up. 

Maybe something like:

“Whether you use Swift packages to quickly add functionality to your project or you choose to create packages of your own, the Swift package ecosystem is a vibrant and growing part of improving your productivity as a developer.”

**Review by dempseyatgithub — COMMENTED** [review]
I really like how the page has turned out!

I do have a few minor items, plus a suggested edit and a suggested closing sentence, instead of ending the page on a list of searches.

File: _data/packages/packages.yml (hunk 9 above)
Comment (daveverwer) [reply _data/packages/packages.yml]: Fixed.

File: _data/packages/packages.yml (hunk 10 above)
Comment (daveverwer) [reply _data/packages/packages.yml]: Fixed

File: _data/packages/packages.yml (hunk 11 above)
Comment (daveverwer) [reply _data/packages/packages.yml]: Fixed

File: packages/index.md (hunk 12 above)
Comment (daveverwer) [reply packages/index.md:16]: Fixed.

File: packages/index.md (hunk 13 above)
Comment (daveverwer) [reply packages/index.md]: Added. Thanks for all your feedback James!

**Comment by daveverwer** [comment]
@swift-ci please test

**Comment by daveverwer** [comment]
@dempseyatgithub I integrated this, but I think it made more sense when the section title was "Next Steps". I had other feedback today to change the heading of this section, and now I think it makes more sense without this final sentence. 

Screenshot 2023-10-31 at 20 51 02@2x

I'm going to remove it and leave the section titled "Advanced Search". Thank you for the feedback though!

**Comment by daveverwer** [comment]
@swift-ci please test

File: packages/index.md (hunk 13 above)
Comment (dempseyatgithub) [reply packages/index.md]: > Added. Thanks for all your feedback James!

You're welcome - thank you for all your work on SPI and these pages Dave!

**Comment by daveverwer** [comment]
@swift-ci please test

**Pull request merged**